src/Symfony/Component/Notifier/Recipient/Assert.php

```diff
@@ -0,0 +1,47 @@
+<?php
+
+/*
+ * This file is part of the Symfony package.
+ *
+ * (c) Fabien Potencier <fabien@symfony.com>
+ *
+ * For the full copyright and license information, please view the LICENSE
+ * file that was distributed with this source code.
+ */
+
+namespace Symfony\Component\Notifier\Recipient;
+
+use Symfony\Component\Notifier\Exception\InvalidArgumentException;
+
+/**
+ * @author Jan Schädlich <jan.schaedlich@sensiolabs.de>
+ *
+ * @experimental in 5.1
+ */
+final class Assert
```

Comment:
I am not so convinced that we need this class. If we want to reuse the validation logic, maybe we should rather simply add internal methods in the `EmailRecipient` and `SmsRecipient` classes.

Reply:
Instead of implementing static validation methods in the `EmailRecipient` and `SmsRecipient` classes, we could also introduce dedicated traits to reuse the validation logic. Something like:

```
trait EmailValidationTrait
{
    private static function ensureEmailIsValid(string $email): void
    {
        if ('' === $email) {
            throw new InvalidArgumentException('The email must be set.');
        }

        if (!preg_match('/^[a-zA-Z0-9.!#$%&\'*+\\/=?^_`{|}~-]+@[a-zA-Z0-9](?:[a-zA-Z0-9-]{0,61}[a-zA-Z0-9])?(?:\.[a-zA-Z0-9](?:[a-zA-Z0-9-]{0,61}[a-zA-Z0-9])?)+$/', $email)) {
            throw new InvalidArgumentException('The email is not valid.');
        }
    }
}
```
```
trait PhoneValidationTrait
{
    private static function ensurePhoneIsValid(string $phone): void
    {
        if ('' === $phone) {
            throw new InvalidArgumentException('The phone number must be set (no spaces, international code like in +3312345678).');
        }

        if (!preg_match('/^\+[0-9]*/', $phone)) {
            throw new InvalidArgumentException('The phone number is not valid (no spaces, international code like in +3312345678).');
        }
    }
}
```

```
class Recipient implements EmailRecipientInterface, SmsRecipientInterface
{
    use EmailValidationTrait;
    use PhoneValidationTrait;

    ...

    public function __construct(string $email, string $phone)
    {
        self::ensureEmailIsValid($email);
        self::ensurePhoneIsValid($phone);

        ...

    }

    ...

}

What do you think?

```

Reply:
@xabbuh WDYT?